packages/core/src/codecatalyst/auth.ts

```diff
@@ -118,6 +155,54 @@ export class CodeCatalystAuthenticationProvider {
         await this.secondaryAuth.restoreConnection()
     }
 
+    private async accessDeniedExceptionHandler() {
+        if (!this.isConnectionValid()) return
+
+        // Check if CodeWhisper and CodeCatalyst share the same connection
+        const isSharedConn =
+            this.secondaryAuth.activeConnection?.id == CodeWhispererAuth.instance.secondaryAuth.activeConnection?.id
+
+        if (isSharedConn) {
+            await codeWhispererClient.listAvailableCustomizations()
+        }
+
+        this.updateConnectionState(this.activeConnection!, { shared: isSharedConn })
+
+        const isPartialExpiration = isSharedConn && !CodeWhispererAuth.instance.isConnectionExpired()
+
+        getLogger().info('CodeCatalyst scopes are expired.')
+
+        this.setScopeExpired(true)
+        await setCodeCatalystConnectedContext(this.isConnectionValid())
+
+        this.showReauthenticationPrompt(this.activeConnection!, isPartialExpiration)
+    }
+
+    public async showReauthenticationPrompt(conn: Connection, isPartialExpiration?: boolean): Promise<void> {
+        const settings = PromptSettings.instance
+        const shouldShow = await settings.isPromptEnabled('codeCatalystConnectionExpired')
+        if (this.suppressNextReauthPrompt) {
+            this.suppressNextReauthPrompt = false
```

Comment:
this logic looks pretty fragile

Reply:
Yes, you're totally right. Passing the value down is a much better idea